### ISSUE TITLE
1.4.0-M2 writes unexpected control characters to shell

### ISSUE THREAD

**Issue by jastice** [sbt/sbt#5784]
## steps

Import an sbt project set to version 1.4.0-M2 into IntelliJ with Scala plugin as sbt project. Do not enable "use shell for imports"

## problem

The import fails to complete. In the log we see what looks like a control character: `[0J`
This apparently breaks the importing - the process just hangs there instead of accepting input.

From the log:
```
/Library/Java/JavaVirtualMachines/jb-jdk-11.0.6-b520.43/Contents/Home/bin/java -Djline.terminal=jline.UnsupportedTerminal -Dsbt.log.noformat=true -Dfile.encoding=UTF-8 -Didea.managed=true -Dfile.encoding=UTF-8 -jar /Users/jast/workspace/scalaUltimate/target/plugin/Scala/launcher/sbt-launch.jar early(addPluginSbtFile="""/private/var/folders/5k/4b5my29x26q_dhrys3ghf9wh0000gn/T/idea7.sbt""")
[info] welcome to sbt 1.4.0-M2 (JetBrains s.r.o Java 11.0.6)
[0J
[info] loading global plugins from /Users/jast/.sbt/1.0/plugins
[0J[info] loading settings for project untitled52-build from idea7.sbt ...
[0J[info] loading project definition from /Users/jast/playspace/untitled52/project
[0J[info] loading settings for project untitled52 from build.sbt ...
[0J[info] set current project to untitled52 (in build file:/Users/jast/playspace/untitled52/)
[0J[info]
[0J[info] Here are some highlights of this release:
[0J[info]   - Build server protocol (BSP) support
[0J[info]   - sbtn: a native thin client for sbt
[0J[info]   - VirtualFile + RemoteCache: caches build artifacts across different machines
[0J[info]   - ThisBuild / versionScheme to take the guessing out of eviction warning
[0J[info] See http://eed3si9n.com/sbt-1.4.0-beta for full release notes.
[0J[info] Hide the banner for this release by running `skipBanner`.
[0J[info] sbt server started at local:///Users/jast/.sbt/1.0/server/a58893d9233954316961/sock
[0J
```

note that the `-Dsbt.log.noformat=true` flag is passed, so I wouldn't expect any control characters.

## notes

The import works fine over the sbt shell integration, which overrides the console prompt.

**Comment by eed3si9n** [comment]
`Esc[0J` :  Clear screen from cursor down | ED0, according to ascii-table.com.

**Comment by eatkins** [comment]
Does it work if you set ~~`-Dsbt.io.virtual=true`~~ `-Dsbt.io.virtual=false`?

**Comment by jastice** [comment]
> Does it work if you set ~`-Dsbt.io.virtual=true`~ `-Dsbt.io.virtual=false`?

It gets even funkier:
```
/Library/Java/JavaVirtualMachines/jdk-9.0.1.jdk/Contents/Home/bin/java -Djline.terminal=jline.UnsupportedTerminal -Dsbt.log.noformat=true -Dfile.encoding=UTF-8 -Xmx2048M -Dsbt.io.virtual=false -Dsbt.supershell=false -Didea.managed=true -Dfile.encoding=UTF-8 -jar /Users/jast/workspace/scalaUltimate/target/plugin/Scala/launcher/sbt-launch.jar
[info] welcome to sbt 1.4.0-M2 (Oracle Corporation Java 9.0.1)
[info] loading global plugins from /Users/jast/.sbt/1.0/plugins
[info] loading project definition from /Users/jast/playspace/untitled52/project
[info] loading settings for project untitled52 from build.sbt ...
[info] set current project to untitled52 (in build file:/Users/jast/playspace/untitled52/)
[info]
[info] Here are some highlights of this release:
[info]   - Build server protocol (BSP) support
[info]   - sbtn: a native thin client for sbt
[info]   - VirtualFile + RemoteCache: caches build artifacts across different machines
[info]   - ThisBuild / versionScheme to take the guessing out of eviction warning
[info] See http://eed3si9n.com/sbt-1.4.0-beta for full release notes.
[info] Hide the banner for this release by running `skipBanner`.
[info] sbt server started at local:///Users/jast/.sbt/1.0/server/a58893d9233954316961/sock
[0J
[?2004h>....
[?2004h>....
[?2004h>....
[?2004h>....
[?2004h>....
[0J
[?2004h>....
[?2004h>....
[?2004h>....
[?2004h>....
[?2004h>....
[0J
[...]
```
(pattern keeps repeating)

**Comment by eatkins** [comment]
I don't think that the control characters are breaking project import though they definitely are annoying. For some reason, the jline 3 linereader wasn't able to read the command from intellij (I'm assuming the intellij is managing a forked sbt process and feeding the command input through the forked process outputstream?). I was able to get build import to work again with 1.4.0-SNAPSHOT by reverting to jline 2 for this use case.

**Comment by eed3si9n** [comment]
> For some reason, the jline 3 linereader wasn't able to read the command from intellij (I'm assuming the intellij is managing a forked sbt process and feeding the command input through the forked process outputstream?)

To put simplistically, is this a bug on JLine 3 or IntelliJ import? If it's JLine 3, maybe we should also report to upstream in some future (like sbt 2.x) when we drop JLine 2.

**Comment by eatkins** [comment]
It's probably some kind of misuse on my end and I would rather not spend more time trying to fix it right now.

**Comment by eatkins** [comment]
But after about 45 minutes of debugging, it became clear that the problem was that the jline 3 implementation LineReader.readLine was not returning a command and when I switched to jline 2, it worked fine. It's definitely unsatisfying but I have gotten very frustrated with debugging jline3.

**Comment by eatkins** [comment]
Is dropping scala 2.12 support on the roadmap for sbt 2? That would be a requirement for dropping jline 2. I hate to say it because the tab completions in jline 3 are so much better than jline 2 but jline 3 is so finicky that I worry about using it exclusively.

**Comment by jastice** [comment]
Thanks for looking into this and the quick fix!
I'm hoping by the time sbt 2.x is ready we'll have a smooth BSP interaction rather than rely on pasting strings into the sbt shell ;)

**Comment by OlegYch** [comment]
same problem on 1.4.0-RC1

**Comment by eatkins** [comment]
@olegych I need a repro. I fixed intellij import on my computer so I don't know what your problem is.

**Comment by OlegYch** [comment]
@eatkins 
the problem is that `-Dsbt.log.noformat=true` still causes control characters in shell and it doesn't accept input, at least on windows:
```
$ set SBT_OPTS=-Dsbt.log.noformat=true && sbt
[info] welcome to sbt 1.4.0-RC1 (Oracle Corporation Java 1.8.0_261)
[info] sbt server started at local:sbt-server-16350a8cf4550c58fe8c
>....
```

**Comment by eatkins** [comment]
Cool. Thanks.

**Comment by eatkins** [comment]
Fixed by #5854
